hs/src/Reach/Eval/Core.hs

```diff
@@ -3124,17 +3124,22 @@ evalPrim p sargs =
       retV $ (lvl, SLV_Bool at $ M.member (bunpack bs) vm)
     SLPrim_Object_set -> do
       at <- withAt id
-      (obj, fieldV, valV) <- three_args
-      (objTy, objDLA) <- compileTypeOf obj
-      (valTy, valDLA) <- compileTypeOf valV
-      fieldName <- bunpack <$> mustBeBytes fieldV
-      case objTy of
+      (obj_v, field_v, val_v) <- three_args
+      (obj_t, obj_a) <- compileTypeOf obj_v
+      (val_t, val_a) <- compileTypeOf val_v
+      fieldName <- bunpack <$> mustBeBytes field_v
+      case obj_t of
         T_Object objFields -> do
-          let objFields' = M.insert fieldName valTy objFields
-          let objTy' = T_Object objFields'
-          let dle = DLE_ObjectSet at objDLA fieldName valDLA
-          dlv <- ctxt_lift_expr (DLVar at Nothing objTy') dle
-          return (lvl, SLV_DLVar dlv)
+          let objFields' = M.insert fieldName val_t objFields
+          let result_t = T_Object objFields'
+          -- DLE_ObjectSet only applies when updating an existing field,
+          -- not when adding a new field or changing a field's type
+          case obj_t == result_t of
+            True -> do
+              let dle = DLE_ObjectSet at obj_a fieldName val_a
+              dlv <- ctxt_lift_expr (DLVar at Nothing result_t) dle
+              return (lvl, SLV_DLVar dlv)
+            False -> undefined -- no idea what parts of this module would pertain to this
```

Comment:
We would go back to the "old" way of doing `Object.set` where you re-construct the object from scratch with a new field. You should replace this `undefined` with `error XXX` and make a test case that has that result if you're having a hard time find one

Reply:
well, the old way is calling Object_set, which does {...o, k: v}; then, that syntax is exploded before even reaching this phase. So, I think I need to do the old way right here. Currently looking at explodeTupleLike to figure that out.

Reply:
You could write `explodingObject_set` in the standard library and then generate a call to the standard library. I think that is better than trying to reimplement what it does here